core/sequencer/election/tracker.go

```diff
@@ -0,0 +1,201 @@
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/monitoring"
+	"github.com/google/trillian/util/election2"
+)
+
+const resourceLabel = "resource"
+
+var (
+	once     sync.Once
+	isMaster monitoring.Gauge
+)
+
+func createMetrics(mf monitoring.MetricFactory) {
+	isMaster = mf.NewGauge(
+		"is_master",
+		"Set to 1 for resources for which this instance is currently master",
+		resourceLabel)
+}
+
+type mastership struct {
+	e        election2.Election
+	acquired time.Time
+}
+
+// Tracker tracks mastership of a collection of resources.
+type Tracker struct {
+	factory     election2.Factory
+	maxHold     time.Duration
+	master      map[string]mastership
+	masterMu    sync.RWMutex
+	watching    map[string]bool
+	watchingMu  sync.RWMutex
+	newResource chan string
+}
+
+// NewTracker returns a new mastership tracker.
+func NewTracker(factory election2.Factory, maxHold time.Duration, metricFactory monitoring.MetricFactory) *Tracker {
+	once.Do(func() { createMetrics(metricFactory) })
+	return &Tracker{
+		factory:     factory,
+		maxHold:     maxHold,
+		master:      make(map[string]mastership),
+		watching:    make(map[string]bool),
+		newResource: make(chan string),
+	}
+}
+
+// AddResource makes the mastership tracker aware of new resources.
+// The same resource may be added an unlimited number of times.
+func (mt *Tracker) AddResource(res string) {
+	mt.newResource <- res
+}
+
+// Run starts new watchers for new resources.
+func (mt *Tracker) Run(ctx context.Context) {
+	for {
+		select {
+		case res := <-mt.newResource:
+			if mt.setWatching(res) {
+				go func() {
+					defer mt.setNotWatching(res)
+					if err := mt.watchResource(ctx, res); err != nil {
+						glog.Errorf("watchResource(%v): %v", res, err)
+					}
+				}()
+			}
+		case <-ctx.Done():
+			glog.Infof("election: Run() exiting due to expired context: %v", ctx.Err())
+			return
+		}
+	}
+}
+
+// watchResource is a blocking method that runs elections for res and updates mt.master.
+func (mt *Tracker) watchResource(ctx context.Context, res string) error {
+	e, err := mt.factory.NewElection(ctx, res)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := e.Close(ctx); err != nil {
+			glog.Warningf("election.Close(%v): %v", res, err)
+		}
+	}()
+
+	for err := error(nil); err == nil; err = ctx.Err() {
+		if err := mt.watchOnce(ctx, e, res); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// watchOnce waits until it acquires mastership, marks itself as master for res,
+// and then waits until either resign duration has passed or it loses
+// mastership, at which point it marks itself as not master for res.  Returns
+// an error if there were problems with acquiring mastership or resigning.
+func (mt *Tracker) watchOnce(ctx context.Context, e election2.Election, res string) error {
+	mt.setNotMaster(res)
```

Comment:
This line looks as a no-op to me, as each resource has only one `watchOnce` at a time which always `setNotMaster` on exit.